src/Todo/DB.hs

```diff
@@ -0,0 +1,54 @@
+module Todo.DB
+    ( TodoDB (..)
+    , DBError (..)
+    , addTodo
+    , addTag
+    , deleteTag
+    , editTag
+    ) where
+
+import Data.HashSet (HashSet)
+import Data.Foldable (toList)
+import qualified Data.HashSet as HSet
+
+import Todo.Model (Todo (..), Tag (..))
+import Data.NESet (NonEmptySet (..))
+import qualified Data.NESet as NESet
+
+data TodoDB = TodoDB
+    { dbTodos :: HashSet Todo
+    , dbTags  :: HashSet Tag
+    } deriving (Eq, Show)
+
+data DBError = SingleTaggedTodo | NotRegisteredTags deriving (Eq, Show)
+
+addTodo :: Todo -> TodoDB -> Either DBError TodoDB
+addTodo todo@Todo {..} TodoDB {..} 
+    | f == mempty = Right TodoDB { dbTodos = HSet.insert todo dbTodos, .. }
+    | otherwise   = Left NotRegisteredTags
+  where
+    f = NESet.filter (\t -> not $ HSet.member t dbTags) todoTags
+
+addTag :: Tag -> TodoDB -> TodoDB
+addTag tag TodoDB {..} = TodoDB { dbTags = HSet.insert tag dbTags, .. }
+
+deleteTag :: Tag -> TodoDB -> Either DBError TodoDB
+deleteTag tag TodoDB {..} = case mapM removeTag $ toList dbTodos of
+    Nothing -> Left SingleTaggedTodo
+    Just x  -> Right TodoDB { dbTodos = HSet.fromList x, ..}
+  where
+    removeTag :: Todo -> Maybe Todo
+    removeTag Todo {..} = 
+        (NESet.fromSet $ NESet.delete tag todoTags) >>= \ts -> Just Todo { todoTags = ts, .. }
```

Comment:
Just note for future. Maybe you will rewrite this in future or maybe not. But I think that this code can be written slightly better (in my opinion) using `do`-notation. I didn't tell about it yet. But you can write this code block like this:

```haskell
removeTag :: Todo -> Maybe Todo
removeTag Todo {..} = do 
  ts <- NESet.fromSet $ NESet.delete tag todoTags
  Just Todo { todoTags = ts, .. }
```

Syntax

```haskell
do
  x <- foo
  smth x
```

is just a syntax sugar for

```haskell
foo >>= \x -> smth x
```